docs/content/any/reference/polar/polar-syntax.md

```diff
@@ -586,3 +605,337 @@ To add an inline query to a Polar file, use the `?=` operator:
+These blocks are especially useful for expressing role-based access control logic.
+
+
+The simplest form of a block looks like this:
+
+```polar
+# Actor block
+actor User {}
+
+# Resource block
+resource Repository {}
+```
+
+In the above example, `User` and `Repository` must be registered application types.
+
+Inside of a block, you can declare [permissions](#permission-declarations), [roles](#role-declarations), and [relations](#relation-declarations) and write [shorthand rules](#shorthand-rules).
+
+A more complete block looks like this:
+
+{{< code codeLang="polar">}}
+resource Repository {
+  ### Permission, role, and relation declarations ###
+  permissions = ["read", "push"];  # permissions that can be granted for the actor or resource
+  roles = ["contributor", "maintainer", "admin"];  # roles that can be assigned for the actor or resource
+  relations = { parent: Organization };  # relations to other actors/resources
+
+  ### Shorthand Rules ###
+  "read" if "contributor"  # "contributor" role grants the "read" permission
+  "push" if "maintainer"  # "member" role grants the "read" permission
+  "maintainer" if "admin"  # "admin" role grants the "maintainer" role
+}
+{{< /code >}}
+<!--
+TODO: should we add the data-linking rules here too? I'm thinking in case
+someone just copies and pastes this whole thing
+-->
+
+Once you have declared a block, you can use the built-in [Actor and Resource
+specializers](#actor-and-resource-specializers) to match all declared actor and
+resource types.
+
+### Permission Declarations
+
+You can specify the permissions that are available for an actor or resource type using the following syntax:
+
+{{< code codeLang="polar" hl_lines="2">}}
+resource Repository {
+  permissions = ["read", "push"];
+}
+{{< /code >}}
+
+Permissions are always Strings. You must declare permissions in order to use them in [shorthand rules](#shorthand-rules).
+
+### Role Declarations
+
+You can specify the roles that are available for an actor or resource type using the following syntax:
+
+{{< code codeLang="polar" hl_lines="2">}}
+resource Repository {
+  roles = ["contributor", "maintainer", "admin"];
+}
+{{< /code >}}
+
+Roles are always Strings. You must declare roles in order to use them in [shorthand rules](#shorthand-rules).
+
+In order to use roles, you must write at least one `has_role` rule that gets
+user-role assignments stored in your application. This rule takes the following
+form:
+
+```polar
+has_role(actor: Actor, name: String, resource: Resource) if ...
+```
+
+For example,
+
+```polar
+# User-role assignment hook - required when using roles
+has_role(user: User, name: String, repo: Repository) if
+  # Look up user-role assignments from application, e.g.
+  role in user.roles and
+  role.repo_id = repo.id;
+```
+
+The `name` argument corresponds to the role names in the declaration list. The
+`has_role` rule must handle every declared role name.
```

Comment:
might be worth explaining the failure mode if the `has_role` rule doesn't handle every declared role name